File: streampark-flink/streampark-flink-connector/streampark-flink-connector-hbase/src/main/scala/org/apache/streampark/flink/connector/hbase/request/HBaseRequest.scala
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.apache.streampark.flink.connector.hbase.request

import org.apache.streampark.common.util.{Logger, Utils}
import org.apache.streampark.flink.connector.hbase.bean.HBaseQuery

import org.apache.flink.api.common.typeinfo.TypeInformation
import org.apache.flink.configuration.Configuration
import org.apache.flink.streaming.api.scala._
import org.apache.flink.streaming.api.scala.async.{ResultFuture, RichAsyncFunction}
import org.apache.hadoop.hbase.client.{Result, ResultScanner, Table}

import java.util.Properties
import java.util.concurrent.{CompletableFuture, Executors, ExecutorService, TimeUnit}
import java.util.function.{Consumer, Supplier}

import scala.annotation.meta.param
import org.apache.streampark.common.util.Implicits._

object HBaseRequest {

  def apply[T: TypeInformation](
      @(transient @param) stream: DataStream[T],
      property: Properties = new Properties()): HBaseRequest[T] =
    new HBaseRequest[T](stream, property)

}

class HBaseRequest[T: TypeInformation](
    @(transient @param) private val stream: DataStream[T],
    property: Properties = new Properties()) {

  def requestOrdered[R: TypeInformation](
      queryFunc: T => HBaseQuery,
      resultFunc: (T, Result) => R,
      timeout: Long = 1000,
      capacity: Int = 10)(implicit prop: Properties): DataStream[R] = {
    Utils.copyProperties(property, prop)
    val async = new HBaseAsyncFunction[T, R](prop, queryFunc, resultFunc, capacity)
    AsyncDataStream.orderedWait(stream, async, timeout, TimeUnit.MILLISECONDS, capacity)
  }

  def requestUnordered[R: TypeInformation](
      queryFunc: T => HBaseQuery,
      resultFunc: (T, Result) => R,
      timeout: Long = 1000,
      capacity: Int = 10)(implicit prop: Properties): DataStream[R] = {
    Utils.copyProperties(property, prop)
    val async = new HBaseAsyncFunction[T, R](prop, queryFunc, resultFunc, capacity)
    AsyncDataStream.unorderedWait(stream, async, timeout, TimeUnit.MILLISECONDS, capacity)
  }

}

class HBaseAsyncFunction[T: TypeInformation, R: TypeInformation](
    prop: Properties,
    queryFunc: T => HBaseQuery,
    resultFunc: (T, Result) => R,
    capacity: Int)
  extends RichAsyncFunction[T, R]
  with Logger {
  @transient private[this] var table: Table = _
  @transient private[this] var executorService: ExecutorService = _

  override def open(parameters: Configuration): Unit = {
    super.open(parameters)
    executorService = Executors.newFixedThreadPool(capacity)
  }

  override def asyncInvoke(input: T, resultFuture: async.ResultFuture[R]): Unit = {
    CompletableFuture
      .supplyAsync(
        new Supplier[ResultScanner]() {
          override def get(): ResultScanner = {
            val query = queryFunc(input)
            require(
              query != null && query.getTable != null,
              "[StreamPark] HBaseRequest query and query's attr table must not be null ")
            table = query.getTable(prop)
            table.getScanner(query)
          }
        },
        executorService
      )
      .thenAccept(new Consumer[ResultScanner] {
        override def accept(result: ResultScanner): Unit = {
          val list = result.toList
          if (list.isEmpty) {
            resultFuture.complete(List(resultFunc(input, Result.EMPTY_RESULT)))
          } else {
            resultFuture.complete(list.map(r => resultFunc(input, r)))
          }
        }
      })
  }

  override def timeout(input: T, resultFuture: ResultFuture[R]): Unit = {
    logWarn("HBaseASync request timeout. retrying... ")
    asyncInvoke(input, resultFuture)
  }

  override def close(): Unit = {
    super.close()
    table.close()
    if (!executorService.isShutdown) {
      executorService.shutdown()
    }
  }
}

File: streampark-flink/streampark-flink-connector/streampark-flink-connector-hbase/src/main/java/org/apache/streampark/flink/connector/hbase/request/HBaseRequest.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.apache.streampark.flink.connector.hbase.request;

import org.apache.streampark.common.util.Utils;
import org.apache.streampark.flink.connector.hbase.bean.HBaseQuery;
import org.apache.streampark.flink.connector.hbase.function.HBaseQueryFunction;
import org.apache.streampark.flink.connector.hbase.function.HBaseResultFunction;

import org.apache.flink.configuration.Configuration;
import org.apache.flink.streaming.api.datastream.AsyncDataStream;
import org.apache.flink.streaming.api.datastream.DataStream;
import org.apache.flink.streaming.api.datastream.SingleOutputStreamOperator;
import org.apache.flink.streaming.api.functions.async.ResultFuture;
import org.apache.flink.streaming.api.functions.async.RichAsyncFunction;
import org.apache.hadoop.hbase.client.Result;
import org.apache.hadoop.hbase.client.ResultScanner;
import org.apache.hadoop.hbase.client.Table;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.IOException;
import java.util.ArrayList;
import java.util.List;
import java.util.Properties;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.TimeUnit;

/** HBase async request helper. */
public class HBaseRequest<T> {

    private final DataStream<T> stream;
    private final Properties property;

    public HBaseRequest(DataStream<T> stream, Properties property) {
        this.stream = stream;
        this.property = property != null ? property : new Properties();
    }

    public static <T> HBaseRequest<T> of(DataStream<T> stream, Properties property) {
        return new HBaseRequest<>(stream, property);
    }

    public <R> SingleOutputStreamOperator<R> requestOrdered(
            HBaseQueryFunction<T> queryFunc,
            HBaseAsyncResultFunction<T, R> resultFunc,
            long timeout,
            int capacity,
            Properties prop) {
        Utils.copyProperties(property, prop);
        HBaseAsyncFunction<T, R> async = new HBaseAsyncFunction<>(prop, queryFunc, resultFunc, capacity);
        return AsyncDataStream.orderedWait(stream, async, timeout, TimeUnit.MILLISECONDS, capacity);
    }

    public <R> SingleOutputStreamOperator<R> requestUnordered(
            HBaseQueryFunction<T> queryFunc,
            HBaseAsyncResultFunction<T, R> resultFunc,
            long timeout,
            int capacity,
            Properties prop) {
        Utils.copyProperties(property, prop);
        HBaseAsyncFunction<T, R> async = new HBaseAsyncFunction<>(prop, queryFunc, resultFunc, capacity);
        return AsyncDataStream.unorderedWait(stream, async, timeout, TimeUnit.MILLISECONDS, capacity);
    }

    @FunctionalInterface
    public interface HBaseAsyncResultFunction<T, R> extends java.io.Serializable {
        R apply(T input, Result result);
    }

    /** HBase async lookup function. */
    public static class HBaseAsyncFunction<T, R> extends RichAsyncFunction<T, R> {

        private static final Logger LOG = LoggerFactory.getLogger(HBaseAsyncFunction.class);

        private final Properties prop;
        private final HBaseQueryFunction<T> queryFunc;
        private final HBaseAsyncResultFunction<T, R> resultFunc;
        private final int capacity;
        private transient Table table;
        private transient ExecutorService executorService;

        public HBaseAsyncFunction(
                Properties prop,
                HBaseQueryFunction<T> queryFunc,
                HBaseAsyncResultFunction<T, R> resultFunc,
                int capacity) {
            this.prop = prop;
            this.queryFunc = queryFunc;
            this.resultFunc = resultFunc;
            this.capacity = capacity;
        }

        @Override
        public void open(Configuration parameters) {
            executorService = Executors.newFixedThreadPool(capacity);
        }

        @Override
        public void asyncInvoke(T input, ResultFuture<R> resultFuture) {
            CompletableFuture.supplyAsync(
                            () -> {
                                HBaseQuery query = queryFunc.query(input);
                                if (query == null || query.getTable() == null) {
                                    throw new IllegalArgumentException(
                                            "[StreamPark] HBaseRequest query and query's attr table must not be null ");
                                }
                                table = query.getTable(prop);
                                try {
                                    return table.getScanner(query);
                                } catch (IOException e) {
                                    throw new RuntimeException(e);
                                }
                            },
                            executorService)
                    .thenAccept(
                            scanner -> {
                                List<R> outputs = new ArrayList<>();
                                boolean hasResult = false;
                                for (Result result : scanner) {
                                    hasResult = true;
                                    outputs.add(resultFunc.apply(input, result));
                                }
                                if (!hasResult) {
                                    outputs.add(resultFunc.apply(input, Result.EMPTY_RESULT));
                                }
                                resultFuture.complete(outputs);
                            });
        }

        @Override
        public void timeout(T input, ResultFuture<R> resultFuture) {
            LOG.warn("HBaseASync request timeout. retrying... ");
            asyncInvoke(input, resultFuture);
        }

        @Override
        public void close() throws Exception {
            if (table != null) {
                table.close();
            }
            if (executorService != null && !executorService.isShutdown()) {
                executorService.shutdown();
            }
        }
    }
}
